src/attacks.rs

```diff
@@ -14,15 +22,214 @@ pub enum DepthReduceSet {
     /// Parameter  is size of the resulting G-S graph desired
     ValiantAB16(usize),
     /// depth of the resulting G-S graph and some specific parameters
-    Greedy(usize, GreedyParams),
+    GreedyDepth(usize, GreedyParams),
+    /// Variation of Greedy attack that has as `target` the resulting size of set S.
+    GreedySize(usize, GreedyParams),
 }
 
 pub fn depth_reduce(g: &mut Graph, drs: DepthReduceSet) -> HashSet<usize> {
     match drs {
         DepthReduceSet::ValiantDepth(_) => valiant_reduce(g, drs),
         DepthReduceSet::ValiantSize(_) => valiant_reduce(g, drs),
         DepthReduceSet::ValiantAB16(_) => valiant_reduce(g, drs),
-        DepthReduceSet::Greedy(target, p) => greedy_reduce(g, target, p),
+        DepthReduceSet::GreedyDepth(_, _) => greedy_reduce(g, drs),
+        DepthReduceSet::GreedySize(_, _) => greedy_reduce(g, drs),
+    }
+}
+
+/// Target of an attack, either the depth should be smaller than `Depth` or
+/// the exclusion set `S` should have a size bigger than `Size` (actually
+/// we want to hit a target as close as possible as those thresholds). All
+/// targets are specified in relation to the size of the graph `G` (not to
+/// be confused with the target size of set `S`).
+// FIXME: Overlapping with `DepthReduceSet` internal values, this should
+//  replace that.
+#[derive(Debug)]
+pub enum AttackTarget {
+    Depth(f64),
+    Size(f64),
+}
+
+/// Range of targets to try (to find the optimum value) from `start`, increasing
+/// by `interval` until `end` is reached or surpassed.
+// FIXME: Using this instead of `std::ops::Range<f64>` because Rust correctly
+//  doesn't allow iterating over floating point values but there is probably
+//  an easier way than coding this from scratch.
+// FIXME: Assert range validity in the struct itself instead of on the caller
+//  (`attack`).
+#[derive(Debug)]
+pub struct TargetRange {
+    pub start: f64,
+    pub interval: f64,
+    pub end: f64,
+}
+
+#[derive(Debug)]
+pub struct AttackProfile {
+    pub runs: usize,
+    pub target: AttackTarget,
+    pub range: TargetRange,
+    pub attack: DepthReduceSet,
```

Comment:
The more I play with the code, the more I'm of the opinion that we should strive for something much simpler in terms of impl. that allows at the same time broader tests cases:
 ```
pub struct AttackProfile {
    pub runs: usize,
    pub attacks: Vec<DepthReduceSet>,
```
One can specify any variations in the depth reducing attacks parameters as well as setting any kind of targets (linear or not, set size or depth) in the same experiment.

Reply:
Of course that can increase the preparation on the user side of the API depending on the experiment. But for the basic case, it is still relatively simple: `(0.15..0.25).step_by(0.01).map(|depth| DepthReduceSet::GreedyDepth( ....)).collect()`